### ISSUE TITLE
CORS multiple issues

### ISSUE THREAD

**Issue by edouardgab** [RESTful-Drupal/restful#1017]
Hi there!

I am starting with the module and trying to understand the different elements. For now, stuck on the login part.

First, the auth passed to the OPTIONS was having an issue, which I correct using 

```
if ($_SERVER['REQUEST_METHOD'] == 'OPTIONS') {
      $this->authenticationManager->setIsOptional(TRUE);
}
```

([https://github.com/RESTful-Drupal/restful/issues/844](https://github.com/RESTful-Drupal/restful/issues/844))

Then, I was getting this: 

> Request header field Authorization is not allowed by Access-Control-Allow-Headers

Which I corrected by adding this to preflight():
```
if ($_SERVER['REQUEST_METHOD'] == 'OPTIONS') {
      $header_bag->add(HttpHeader::create('Access-Control-Allow-Headers', "Content-Type, Authorization"));
}
```
Now, I can almost login, it returns me the token, but the browser still returns me this error:

> No 'Access-Control-Allow-Origin' header is present on the requested resource. Origin 'http://localhost:8080' is therefore not allowed access.

From what I see the the response request, the Access-Control-Allow-Origin is here in the first OPTIONS request, but not there during the GET. Is there an issue here of it is me missing something?

Thanks!

**Comment by edouardgab** [comment]
I was able to correct this by modifying `restful_menu_process_callback` in resful.module.

**From** 
```
$plugin_definition = $resource->getPluginDefinition();
if (!empty($plugin_definition['allowOrigin'])) {
  $response_headers->append(HttpHeader::create('Access-Control-Allow-Origin', $plugin_definition['allowOrigin']));
}
```
**to**
```
$plugin_definition = $resource->getPluginDefinition();
$allowed_origin = empty($plugin_definition['allowOrigin']) ? variable_get('restful_allowed_origin', NULL) : $plugin_definition['allowOrigin'];
if ($allowed_origin) {
  $response_headers->append(HttpHeader::create('Access-Control-Allow-Origin', $allowed_origin));
}
```

Basically using the same method to retrieve the allowOrigin than the one used in Resource.php preflight().

I hope this was the right way to go, please let me know if I am missing something.

Thanks!

**Comment by edouardgab** [comment]
Actually, I am re-openning this so if anyone see this, he could give me a feedback.

**Comment by edouardgab** [comment]
Something else and I am not sure what to do with it. When requesting a resource and not sending the token, it throws:

```
throw new UnauthorizedException('Bad credentials. Anonymous user resolved for a resource that requires authentication.');
```

Which makes it bypass the `restful_menu_process_callback` where the Access-Control-Allow-Origin is set. so the browser returns: 
`No 'Access-Control-Allow-Origin' header is present on the requested resource. Origin 'http://localhost:8080' is therefore not allowed access. The response had HTTP status code 401.`

It doesn't seems like a major issue, but it this the proper way to handle this? 

Thanks for the help!

**Comment by guillaumev** [comment]
Just to say that I'm having the same issue with Access-Control-Allow-Headers. My Access-Control-Allow-Origin is working however, but I had to add allowOrigin = "*" in my plugin definition.

**Comment by jsacksick** [comment]
restful_menu_process_callback() needs to be modified indeed, in order not to hack the module, I did it with a hook_restful_resource_alter().

**Comment by e0ipso** [comment]
> I had to add allowOrigin = "*" in my plugin definition.

I believe this is the proper fix.

**Comment by jsacksick** [comment]
I don't think this is the proper fix as this would allow any origin... The code in restful_menu_process_callback is not consistent with the code in the preflight() method which both check for the plugin definition and the restful_allowed_origin variable.

I believe the code should be modified to check both in restful_menu_process_callback as well...

I implemented the following workaround in the meantime:

```php
/**
 * Implements hook_restful_resource_alter().
 */
function MY_MODULE_restful_resource_alter(ResourceInterface &$resource) {
  $plugin_definition = $resource->getPluginDefinition();
  // restful_menu_process_callback doesn't properly set the allowed origin
  // header, it checks only for the allowedOrigin key in the plugin definition
  // and doesn't take in account the "restful_allowed_origin" variable, so we
  // need to fix that.
  if (!empty($plugin_definition['allowOrigin'])) {
    return;
  }
  $allowed_origin = variable_get('restful_allowed_origin', NULL);
  if (!empty($allowed_origin)) {
    $plugin_definition['allowOrigin'] = $allowed_origin;
    $resource->setPluginDefinition($plugin_definition);
  }
}
```

Additionally, I realized several minutes ago that when doing a DELETE request, the Access-Control-Allow-Origin, as well as other headers aren't sent in the response...

That's due to the fact that the remove method in the Resource plugin doesn't return anything.

As a consequence, the first check restful_delivery() doesn't pass since $var is NULL, so $response->send() isn't called...

I have a workaround for that as well, but it'd be nice to find a proper fix for that scenario....
